In [1]:
from MySQLdb import connect
import pandas as pd

In [3]:
# Connect to your MySQLdb
conn = connect(host='localhost', user='root')
#conn = connect(host='127.0.0.1', user='root', password = 'root', port=8890)

In [4]:
# Open a cursor to perform database operations
cur = conn.cursor()

### SQL Wildcard Characters
- A wildcard character is used to substitute one or more characters in a string.
- Wildcard characters are used with the LIKE operator. The LIKE operator is used in a WHERE clause to search for a specified pattern in a column.

In [5]:
# Return all customers that starts with the letter 'a':
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE 'a%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID                        CustomerName     ContactName  \
0           1                 Alfreds Futterkiste    Maria Anders   
1           2  Ana Trujillo Emparedados y helados    Ana Trujillo   
2           3             Antonio Moreno Taquería  Antonio Moreno   
3           4                     Around the Horn    Thomas Hardy   

                         Address         City PostalCode  Country  
0                  Obere Str. 57       Berlin      12209  Germany  
1  Avda. de la Constitución 2222  México D.F.      05021   Mexico  
2                 Mataderos 2312  México D.F.      05023   Mexico  
3                120 Hanover Sq.       London    WA1 1DP       UK

Wildcard Characters

| Symbol | Description |
| :--- | ---: |
% | Represents zero or more characters
_ | Represents a single character
[] | Represents any single character within the brackets *
^ | Represents any character not in the brackets * |
|- | Represents any single character within the specified range * |
{} | Represents any escaped character **

* Not supported in PostgreSQL and MySQL databases.

** Supported only in Oracle databases.


Using the % Wildcard:
- The % wildcard represents any number of characters, even zero characters.

In [6]:
# Return all customers that ends with the pattern 'es':
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE '%es';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID         CustomerName        ContactName  \
0          11        B's Beverages  Victoria Ashworth   
1          23    Folies gourmandes      Martine Rancé   
2          31  Gourmet Lanchonetes      André Fonseca   
3          34        Hanari Carnes       Mario Pontes   
4          47     LINO-Delicateses   Felipe Izquierdo   
5          57    Paris spécialités     Marie Bertrand   
6          76     Suprêmes délices   Pascale Cartrain   

                    Address             City PostalCode    Country  
0         Fauntleroy Circus           London    EC2 5NT         UK  
1  184, chaussée de Tournai            Lille      59000     France  
2           Av. Brasil, 442         Campinas  04876-786     Brazil  
3           Rua do Paço, 67   Rio de Janeiro  05454-876     Brazil  
4   Ave. 5 de Mayo Porlamar  I. de Margarita       4980  Venezuela  
5   265, boulevard Charonne            Paris      75012     France  
6      Boulevard Tirou, 255        Charleroi     B-6000    Belgium

In [7]:
# Return all customers that contains the pattern 'mer':
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE '%mer%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID                CustomerName        ContactName  \
0          13  Centro comercial Moctezuma    Francisco Chang   
1          15            Comércio Mineiro       Pedro Afonso   
2          46           LILA-Supermercado    Carlos González   
3          51              Mère Paillarde     Jean Fresnière   
4          69            Romero y tomillo   Alejandra Camino   
5          81      Tradição Hipermercados  Anabela Domingues   

                                          Address          City PostalCode  \
0                         Sierras de Granada 9993   México D.F.      05022   
1                            Av. dos Lusíadas, 23     São Paulo  05432-043   
2  Carrera 52 con Ave. Bolívar #65-98 Llano Largo  Barquisimeto       3508   
3                              43 rue St. Laurent      Montréal    H1J 1C3   
4                                     Gran Vía, 1        Madrid      28001   
5                         Av. Inês de Castro, 414     São Paulo  05634-030   

     Country  
0     Mexico  
1     Brazil  
2  Venezuela  
3     Canada  
4      Spain  
5     Brazil

Using the _ Wildcard:
- The _ wildcard represents a single character.
- It can be any character or number, but each _ represents one, and only one, character.

In [8]:
# Return all customers with a City starting with any character, followed by "ondon":
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE City LIKE '_ondon';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID           CustomerName        ContactName  \
0           4        Around the Horn       Thomas Hardy   
1          11          B's Beverages  Victoria Ashworth   
2          16  Consolidated Holdings    Elizabeth Brown   
3          19     Eastern Connection          Ann Devon   
4          53            North/South     Simon Crowther   
5          72     Seven Seas Imports         Hari Kumar   

                        Address    City PostalCode Country  
0               120 Hanover Sq.  London    WA1 1DP      UK  
1             Fauntleroy Circus  London    EC2 5NT      UK  
2   Berkeley Gardens 12 Brewery  London    WX1 6LT      UK  
3                35 King George  London    WX3 6FW      UK  
4  South House 300 Queensbridge  London    SW7 1RZ      UK  
5               90 Wadhurst Rd.  London   OX15 4NB      UK

In [9]:
# Return all customers with a City starting with "L", followed by any 3 characters, ending with "on":
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE City LIKE 'L___on';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID           CustomerName        ContactName  \
0           4        Around the Horn       Thomas Hardy   
1          11          B's Beverages  Victoria Ashworth   
2          16  Consolidated Holdings    Elizabeth Brown   
3          19     Eastern Connection          Ann Devon   
4          53            North/South     Simon Crowther   
5          72     Seven Seas Imports         Hari Kumar   

                        Address    City PostalCode Country  
0               120 Hanover Sq.  London    WA1 1DP      UK  
1             Fauntleroy Circus  London    EC2 5NT      UK  
2   Berkeley Gardens 12 Brewery  London    WX1 6LT      UK  
3                35 King George  London    WX3 6FW      UK  
4  South House 300 Queensbridge  London    SW7 1RZ      UK  
5               90 Wadhurst Rd.  London   OX15 4NB      UK

Using the [] Wildcard:
- The [] wildcard returns a result if any of the characters inside gets a match.

In [10]:
# Return all customers starting with either "b", "s", or "p":
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE '[bsp]%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

ValueError: Length mismatch: Expected axis has 0 elements, new values have 7 elements

Using the - Wildcard:
- The - wildcard allows you to specify a range of characters inside the [] wildcard.

In [12]:
# Return all customers starting with "a", "b", "c", "d", "e" or "f":
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE '[a-f]%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

ValueError: Length mismatch: Expected axis has 0 elements, new values have 7 elements

Combine Wildcards:
- Any wildcard, like % and _ , can be used in combination with other wildcards.

In [13]:
# Return all customers that starts with "a" and are at least 3 characters in length:
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE 'a__%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID                        CustomerName     ContactName  \
0           1                 Alfreds Futterkiste    Maria Anders   
1           2  Ana Trujillo Emparedados y helados    Ana Trujillo   
2           3             Antonio Moreno Taquería  Antonio Moreno   
3           4                     Around the Horn    Thomas Hardy   

                         Address         City PostalCode  Country  
0                  Obere Str. 57       Berlin      12209  Germany  
1  Avda. de la Constitución 2222  México D.F.      05021   Mexico  
2                 Mataderos 2312  México D.F.      05023   Mexico  
3                120 Hanover Sq.       London    WA1 1DP       UK

In [14]:
# Return all customers that have "r" in the second position:
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE CustomerName LIKE '_r%';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID                       CustomerName        ContactName  \
0            4                    Around the Horn       Thomas Hardy   
1           17          Drachenblut Delikatessend       Sven Ottlieb   
2           20                       Ernst Handel      Roland Mendel   
3           25                     Frankenversand      Peter Franken   
4           26                France restauration     Carine Schmitt   
5           27                     Franchi S.p.A.      Paolo Accorti   
6           32            Great Lakes Food Market      Howard Snyder   
7           33               GROSELLA-Restaurante     Manuel Pereira   
8           60            Princesa Isabel Vinhoss   Isabel de Castro   
9           81             Tradição Hipermercados  Anabela Domingues   
10          82  Trail's Head Gourmet Provisioners     Helvetius Nagy   

                      Address       City PostalCode    Country  
0             120 Hanover Sq.     London    WA1 1DP         UK  
1                Walserweg 21     Aachen      52066    Germany  
2                Kirchgasse 6       Graz       8010    Austria  
3           Berliner Platz 43    München      80805    Germany  
4              54, rue Royale     Nantes      44000     France  
5         Via Monte Bianco 34     Torino      10100      Italy  
6            2732 Baker Blvd.     Eugene      97403        USA  
7   5ª Ave. Los Palos Grandes    Caracas       1081  Venezuela  
8      Estrada da saúde n. 58     Lisboa       1756   Portugal  
9     Av. Inês de Castro, 414  São Paulo  05634-030     Brazil  
10          722 DaVinci Blvd.   Kirkland      98034        USA

Without Wildcard:
- If no wildcard is specified, the phrase has to have an exact match to return a result.

In [15]:
# Return all customers from Spain:
cur.execute('''SELECT * FROM sqlw3s.Customers
WHERE Country LIKE 'Spain';''')
df = pd.DataFrame(tuple(t) for t in cur.fetchall())
df.columns = [i[0] for i in cur.description]
df

CustomerID                          CustomerName        ContactName  \
0           8             Bólido Comidas preparadas      Martín Sommer   
1          22  FISSA Fabrica Inter. Salchichas S.A.         Diego Roel   
2          29                Galería del gastrónomo   Eduardo Saavedra   
3          30                   Godos Cocina Típica  José Pedro Freyre   
4          69                      Romero y tomillo   Alejandra Camino   

                  Address       City PostalCode Country  
0          C/ Araquil, 67     Madrid      28023   Spain  
1      C/ Moralzarzal, 86     Madrid      28034   Spain  
2  Rambla de Cataluña, 23  Barcelona      08022   Spain  
3           C/ Romero, 33    Sevilla      41101   Spain  
4             Gran Vía, 1     Madrid      28001   Spain

Microsoft Access Wildcards
- The Microsoft Access Database has some other wildcards:

| Symbol | Description | Example |
| :--- | :---: | ---: |
| * | Represents zero or more characters | bl* finds bl, black, blue, and blob
| ? | Represents a single character | h?t finds hot, hat, and hit
| [] | Represents any single character within the brackets | h[oa]t finds hot and hat, but not hit
| ! | Represents any character not in the brackets | h[!oa]t finds hit, but not hot and hat
| - | Represents any single character within the specified range | c[a-b]t finds cat and cbt
| # | Represents any single numeric character | 2#5 finds 205, 215, 225, 235, 245, 255, 265, 275, 285, and 295
